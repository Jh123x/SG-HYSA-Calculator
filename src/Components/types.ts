import Profile from "../types/profile.ts";

export interface InputArg<Type> {
    label: Readonly<string>;
    tooltip: Readonly<string>;
    fn: (profile: Profile, value: Type) => Profile;
    getStateFromProfile: (profile: Profile) => Type;
}

export interface Field<Type> {
    label: string
    onChange: (v: Type | "") => void
    value: Type | ""
    tooltip?: string
}