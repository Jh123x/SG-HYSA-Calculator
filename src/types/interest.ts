export interface CutoffInterest {
    Cutoff: number
    InterestRatePercent: number
}

export interface Interest {
    cutoffs: Array<CutoffInterest>
    baseRatePercent: number
}
