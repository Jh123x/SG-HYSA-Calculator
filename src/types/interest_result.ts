export class ResultInterest {
    private yearly_interest: number
    private savings: number
    constructor(yearly: number, savings: number) {
        this.yearly_interest = parseFloat(yearly.toFixed(2)) ?? 0;
        this.savings = savings
    }

    toYearlyPercent(): number {
        if (this.savings === 0) return 0

        return (this.yearly_interest / this.savings * 100);
    }

    toYearly(): number {
        return this.yearly_interest
    }

    toMonthly(): number {
        return this.yearly_interest / 12
    }

    toDaily(): number {
        return this.yearly_interest / 365
    }
}