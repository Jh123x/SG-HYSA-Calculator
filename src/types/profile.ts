export default interface Profile {
    Savings: number
    Age: number
    Salary: number
    Spending: number
    Investment: number
    Insurance: number
    GiroTransactions: number
    MonthlyAccIncrease: number
    LoanInstallment: number
    OneTimeLoan: number
    IsNTUCMember: boolean
}

export const NewProfile = ({
    Savings = 0,
    Age = 0,
    Salary = 0,
    Spending = 0,
    Investment = 0,
    Insurance = 0,
    GiroTransactions = 0,
    MonthlyAccIncrease = 0,
    LoanInstallment = 0,
    OneTimeLoan = 0,
    IsNTUCMember = false,
}): Profile => {
    return {
        Savings,
        Age,
        Salary,
        Spending,
        Investment,
        Insurance,
        GiroTransactions,
        MonthlyAccIncrease,
        LoanInstallment,
        IsNTUCMember,
        OneTimeLoan,
    }
}
