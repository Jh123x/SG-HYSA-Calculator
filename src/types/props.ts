import { ResultInterest } from "./interest_result"

export interface ResultProp {
    interest: ResultInterest,
    url: string
    remarks?: string
    lastUpdated: Date
}