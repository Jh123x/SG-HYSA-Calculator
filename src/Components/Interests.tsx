import React, { ReactElement } from "react"
import { Container, Link, Table, TableBody, TableCell, TableContainer, TableHead, TableRow, Typography } from "@mui/material"
import { ResultProp } from "../types/props"
import { primaryColor, bgColor, textColor } from "../consts/colors.ts"
import Profile from "../types/profile.ts"
import { bankInfo } from "../logic/constants.tsx"

export const Result = ({ profile }: { profile: Profile }) => {
    const results: Record<string, ResultProp> = {}
    for (const [name, info] of Object.entries(bankInfo)) {
        results[name] = {
            interest: info.interestFn(profile),
            url: info.url,
            remarks: info.remarks,
            lastUpdated: info.lastUpdated,
        }
    }

    return <Container sx={{ color: primaryColor, width: "100%", maxWidth: "100%" }}>
        <TableContainer>
            <Table>
                <TableHead>
                    <ThemedTableRow>
                        <ThemedTableCell>Account Name</ThemedTableCell>
                        <ThemedTableCell>Yearly Interest</ThemedTableCell>
                        <ThemedTableCell><>Effective Interest<br />Rate (EIR)</></ThemedTableCell>
                        <ThemedTableCell>Webpage</ThemedTableCell>
                        <ThemedTableCell>Remarks</ThemedTableCell>
                        <ThemedTableCell>Updated at</ThemedTableCell>
                    </ThemedTableRow>
                </TableHead>
                <TableBody>
                    {Object.entries(results).map(([bankName, interest]) => displayResult(bankName, interest))}
                </TableBody>
            </Table>
        </TableContainer>
        <Typography variant="caption" sx={{ color: textColor, margin: "10px 0px" }}>
            * Interest rates on their respective websites are subject to change without notice
        </Typography>
    </Container>
}

const displayResult = (bankName: string, info: ResultProp) => {
    const lastUpdated = info.lastUpdated
    return <ThemedTableRow key={bankName}>
        <ThemedTableCell>{bankName}</ThemedTableCell>
        <ThemedTableCell>{info.interest.toYearly() ?? 0}</ThemedTableCell>
        <ThemedTableCell>{info.interest.toYearlyPercent().toFixed(2) ?? ""}</ThemedTableCell>
        <ThemedTableCell>
            <Link href={info.url} target="_blank" sx={{ color: textColor }}>Official Website</Link>
        </ThemedTableCell>
        <ThemedTableCell>{info.remarks}</ThemedTableCell>
        <ThemedTableCell>{lastUpdated}</ThemedTableCell>
    </ThemedTableRow>
}

const ThemedTableRow = ({ children, bankName }: {
    children: Array<ReactElement>
    bankName?: string
}) => <TableRow key={bankName} sx={{ color: textColor, backgroundColor: bgColor }}>{children}</TableRow>

const ThemedTableCell = ({ children }: {
    children: string | number | ReactElement
}) => <TableCell sx={{ color: textColor, backgroundColor: bgColor }}>{children}</TableCell>