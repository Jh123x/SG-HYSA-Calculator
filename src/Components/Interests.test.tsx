import React from "react"
import { Result } from "./Interests"
import { ResultInterest } from "../types/interest_result"

describe("Result Table", () => {
    it("should match snapshot when results is empty", () => {
        expect(<Result results={{}}></Result>).toMatchSnapshot()
    })

    it("should match snapshot when results are full", () => {
        expect(
            <Result results={{
                "UOB Bank": {
                    interest: new ResultInterest(1000),
                    url: "https://jh123x.com",
                    remarks: "Testing remarks"
                }
            }}></Result>
        ).toMatchSnapshot()
    })
})