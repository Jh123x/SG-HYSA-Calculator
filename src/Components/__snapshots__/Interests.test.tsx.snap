// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result Table should match snapshot when results are full 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiTableContainer-root css-t6756l-MuiTableContainer-root"
      >
        <table
          class="MuiTable-root css-1xwxv7r-MuiTable-root"
        >
          <thead
            class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head css-1r57mmw-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-j77fgu-MuiTableCell-root"
                scope="col"
              >
                Bank Name
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-j77fgu-MuiTableCell-root"
                scope="col"
              >
                Yearly Interest
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-j77fgu-MuiTableCell-root"
                scope="col"
              >
                Remarks
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-j77fgu-MuiTableCell-root"
                scope="col"
              >
                Webpage
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root css-1r57mmw-MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1mmjt0s-MuiTableCell-root"
              >
                UOB Bank
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1mmjt0s-MuiTableCell-root"
              >
                1000
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1mmjt0s-MuiTableCell-root"
              >
                Testing remarks
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1mmjt0s-MuiTableCell-root"
              >
                <a
                  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1yzuh6z-MuiTypography-root-MuiLink-root"
                  href="https://jh123x.com"
                  target="_blank"
                >
                  Official Website
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiTableContainer-root css-t6756l-MuiTableContainer-root"
    >
      <table
        class="MuiTable-root css-1xwxv7r-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1r57mmw-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-j77fgu-MuiTableCell-root"
              scope="col"
            >
              Bank Name
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-j77fgu-MuiTableCell-root"
              scope="col"
            >
              Yearly Interest
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-j77fgu-MuiTableCell-root"
              scope="col"
            >
              Remarks
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-j77fgu-MuiTableCell-root"
              scope="col"
            >
              Webpage
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-1r57mmw-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1mmjt0s-MuiTableCell-root"
            >
              UOB Bank
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1mmjt0s-MuiTableCell-root"
            >
              1000
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1mmjt0s-MuiTableCell-root"
            >
              Testing remarks
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1mmjt0s-MuiTableCell-root"
            >
              <a
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1yzuh6z-MuiTypography-root-MuiLink-root"
                href="https://jh123x.com"
                target="_blank"
              >
                Official Website
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Result Table should match snapshot when results is empty 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiTableContainer-root css-t6756l-MuiTableContainer-root"
      >
        <table
          class="MuiTable-root css-1xwxv7r-MuiTable-root"
        >
          <thead
            class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head css-1r57mmw-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-j77fgu-MuiTableCell-root"
                scope="col"
              >
                Bank Name
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-j77fgu-MuiTableCell-root"
                scope="col"
              >
                Yearly Interest
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-j77fgu-MuiTableCell-root"
                scope="col"
              >
                Remarks
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-j77fgu-MuiTableCell-root"
                scope="col"
              >
                Webpage
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
          />
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiTableContainer-root css-t6756l-MuiTableContainer-root"
    >
      <table
        class="MuiTable-root css-1xwxv7r-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1r57mmw-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-j77fgu-MuiTableCell-root"
              scope="col"
            >
              Bank Name
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-j77fgu-MuiTableCell-root"
              scope="col"
            >
              Yearly Interest
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-j77fgu-MuiTableCell-root"
              scope="col"
            >
              Remarks
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-j77fgu-MuiTableCell-root"
              scope="col"
            >
              Webpage
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
        />
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
